### PR TITLE
providers/vmware: Process guestinfo.metadata netplan configuration

### PR DESCRIPTION
**Pull request by pothos**
The network environment can be dynamic and thus needs to be provided as
    VM metadata. Since the format should not depend on whether the VM runs
    uses Ignition and Afterburn or Cloud-Init, the idea is to also support
    the guestinfo.metadata variable as used by Cloud-Init which contains
    Netplan YAML network configuration.
    Add a new command to write out netplan configs to a given directory,
    similar as we do with networkd units. While this is currently just used
    for VMware, other providers could also construct the netplan data type
    to provide netplan configurations if the OS rather wants to use
    NetworkManager than systemd-networkd. For backwards compatibility and
    to not need netplan it would be nice to keep the systemd-networkd
    support as long as its used.

What do you think, rather only write out to `/etc/netplan/afterburn.yaml` and rely on the netplan implementation, or do the processing YAML directly in afterburn?

References:
* https://cloudinit.readthedocs.io/en/latest/reference/datasources/vmware.html#walkthrough-of-guestinfo-keys-transport 
* https://cloudinit.readthedocs.io/en/latest/reference/network-config-format-v2.html
* https://netplan.io/reference/
* https://linux-on-z.blogspot.com/p/using-netplan-on-ibm-z.html

### CODE REVIEW

**Comment by pothos** [comment]
@bgilbert @lucab Any preference? While I think that doing all processing in Afterburn would be nice, the more practical approach would be to write the netplan config file out and let the distro handle it through the regular `netplan` implementation.
Edit: Also an option, don't do it in afterburn but call `vmtoolsd --cmd 'info-get guestinfo.metadata'` directly to write out this file for netplan.

**Comment by bgilbert** [comment]
Thanks for the PR.  Afterburn already has a mechanism for this on VMware, #404, but it isn't very user-friendly (initrd `ip=` syntax is not great).  Afterburn's network configuration support hasn't gotten a lot of attention in recent years, due to the longstanding problem that it's designed for networkd and doesn't map well onto NetworkManager.  I wasn't aware of Netplan and wish I had known about it sooner; it looks like a good fit for Afterburn's needs.

I see an opportunity here to revisit our network configuration design, and reuse existing abstractions rather than maintaining our own. Rather than having the DigitalOcean and Packet providers populate networkd-format structs, they (and the VMware provider) could populate a `netplan-types` struct, and then Afterburn could write that out via a new `--netplan-conf <directory>` option.  The distro would be responsible for configuring Netplan to translate to networkd or NetworkManager as desired.

WDYT?  If you have an immediate need for this, it could be reasonable to add some stub Netplan support for VMware but defer migration of the other providers to a later PR.

**Comment by pothos** [comment]
Thanks for the feedback, ok, I'm actually happy that we generate networkd units directly in afterburn but it's good to know that you are looking for a solution that also covers NetworkManager-only systems. Flatcar doesn't ship netplan (and ok course also doesn't ship [nmstate](https://nmstate.io/)). We could add it to the VMware Flatcar images and limit the use to it at first, and later see if that's something we could use for other providers, too. I'll drop the unit generation then and only parse the file once and emit it again to have a way where other providers could expose a netplan struct as you said.

**Comment by pothos** [comment]
Done but keeping it it draft mode until I test it on a real system

File: src/providers/vmware/amd64.rs
Comment (akutz) [inline src/providers/vmware/amd64.rs:152]: I am not familiar with Rust, so please pardon my ignorance, but does the parsing code care if the `network` key is not the only top-level value in the metadata? Because it would not have to be. Also, we support sending the metadata either as YAML _or_ JSON, so is it possible to support both?

Comment (pothos) [reply src/providers/vmware/amd64.rs:152]: Other keys can be present, but are ignored. Yes, JSON support could be added, too.

Comment (pothos) [reply src/providers/vmware/amd64.rs:152]: Pushed JSON support, please have a look

File: src/providers/vmware/amd64.rs
Comment (pothos) [inline src/providers/vmware/amd64.rs]: Multiple calls on a channel don't work:
```
    2: vmware backdoor error: erpc send_command_len failed
```
and the channel holds the mutable reference, hence to create a new channel, I needed to drop. I think the explicit call is better than doing this implicitly through blocks because this will be confusing for the next person to touch.

**Comment by pothos** [comment]
Tested with a Flatcar VM as follows, the VM didn't have the guestinfo.metadata set on provisioning:
```
cargo build
scp -C target/debug/afterburn core@ESXIVM:
ssh core@ESXIVM
# now in the VM:
./afterburn multi --netplan-configs $PWD/ --provider vmware
# no afterburn.yaml will be written to PWD
/usr/share/oem/bin/vmtoolsd --cmd "info-set guestinfo.metadata bmV0d29yazoKICB2ZXJzaW9uOiAyCiAgZXRoZXJuZXRzOgogICAgbmljczoKICAgICAgbWF0Y2g6CiAgICAgICAgbmFtZTogZW5zKgogICAgICBkaGNwNDogdHJ1ZQo="
/usr/share/oem/bin/vmtoolsd --cmd "info-set guestinfo.metadata.encoding base64"
./afterburn multi --netplan-configs $PWD/ --provider vmware
# the afterburn.yaml gets written to PWD
cat afterburn.yaml
# prints this:
# network:
#  version: 2
#  ethernets:
#    nics:
#      match:
#        name: ens*
#      dhcp4: true
```

File: src/providers/vmware/amd64.rs
Comment (akutz) [reply src/providers/vmware/amd64.rs:152]: Lgtm, thanks!

File: src/providers/vmware/amd64.rs
Comment (akutz) [inline src/providers/vmware/amd64.rs]: Please note there are multiple, documented options for expressing the encoding:

* b64
* base64
* gz+b64
* gzip+base64

**Review by akutz — CHANGES_REQUESTED** [review]
One minor change requested, and then it's good to go!

**Comment by bgilbert** [comment]
CI failures are a failed lint, missing release notes in `docs/release-notes.md`, and the fact that `netplan-types` isn't packaged in Fedora.  We have a rule that new dependencies must be packaged in Fedora first, so we don't have an unknown amount of packaging to do (including transitive dependencies) at release time.  You're not necessarily responsible for handling that yourself (unless you want to!) but it'll need to be addressed.

I won't be able to do the review here, unfortunately, but hopefully one of the other folks on the team can take a look.  Thanks for your work on this PR!

File: Cargo.toml
Comment (bgilbert) [inline Cargo.toml]: Please truncate these to `0.21` etc., so Dependabot doesn't feel the need to bump the Z version.

**Comment by cgwalters** [comment]
>  The distro would be responsible for configuring Netplan to translate to networkd or NetworkManager as desired.

Would flatcar ship code to translate netplan to systemd-networkd?

I am not totally sure we want to own code doing this for NetworkManager.

(Not to architecture astronaut this too much but maybe what could be tractable is translating some networkd configs to NetworkManager...)

**Comment by jepio** [comment]
> > The distro would be responsible for configuring Netplan to translate to networkd or NetworkManager as desired.
> 
> Would flatcar ship code to translate netplan to systemd-networkd?
> 
> I am not totally sure we want to own code doing this for NetworkManager.
> 
> (Not to architecture astronaut this too much but maybe what could be tractable is translating some networkd configs to NetworkManager...)

The plan is like @bgilbert mentioned above "The distro would be responsible for configuring Netplan to translate to networkd or NetworkManager as desired.". There are also plans to write a rust implementation of netplan from what I heard.

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs]: Ok, added that

File: Cargo.toml
Comment (pothos) [reply Cargo.toml]: Ok, done for the others, too

**Comment by pothos** [comment]
I've addressed the partly review, it's ready for a full review.
Can someone do that?
Is there someone I could ask about Rust Fedora packaging?

> Would flatcar ship code to translate netplan to systemd-networkd?
> 
> I am not totally sure we want to own code doing this for NetworkManager.
> 
> (Not to architecture astronaut this too much but maybe what could be tractable is translating some networkd configs to NetworkManager...)

Netplan supports translation to either networkd or NetworkManager. For Flatcar we would add Netplan to the VMware OEM image and invoke it for translation to ephemeral networkd units under `/run` (probably from the initrd to have networking for Ignition, and this would also be carried over to the final system).

File: docs/release-notes.md
Comment (jlebon) [inline docs/release-notes.md]: These notes are usually brief. E.g.

```suggestion
- Add Netplan support on VMware
```

Can you provide usage details in a new `docs/usage/` page instead? See for example `initrd-network-cmdline.md`, which gets rendered here: https://coreos.github.io/afterburn/usage/initrd-network-cmdline/.

File: Cargo.toml
Comment (jlebon) [inline Cargo.toml]: One way to avoid having to package this in Fedora is to gate it behind a feature for now. If we do that though, we should also make sure that CI here is still exercising that path.

That said, you're definitely welcome to package it on Fedora if you're interested! See [this page](https://docs.fedoraproject.org/en-US/package-maintainers/Joining_the_Package_Maintainers/) for details on how to get started.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs:15]: Hmm, I'm unsure about the use of `guestinfo.metadata`. The PR description says:

> Since the format should not depend on whether the VM runs
uses Ignition and Afterburn or Cloud-Init, the idea is to also support
the guestinfo.metadata variable as used by Cloud-Init which contains
Netplan YAML network configuration.

But in the cloud-init case, users would normally put their netplan configs under e.g. `write_files` or something, right? I'm worried re-using the same key as cloud-init on systems with Ignition and Afterburn where one would've expected an Ignition config to live there instead. (Another way to say this: either that key should be read by Ignition or not at all.) I'm not sure on the historical reason why Ignition created its own key rather than conform to the established `guestinfo.metadata` key, but at this point I think it'd add more to the confusion to introduce a reader.

If the goal is to try to land on something that could potentially be implemented cross-provisioning tool, then let's create a new key. E.g. `guestinfo.netplan.metadata`? Or maybe `guestinfo.networking.metadata` to make it agnostic to the networking tool too since one could imagine feeding nmstate configs through there instead. Otherwise, something Afterburn-specific is also fine, like `guestinfo.afterburn.netplan-configs`.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [reply src/providers/vmware/amd64.rs]: Hmm, possibly there's a bug there in `vmw_backdoor` that we need to fix?

> I think the explicit call is better than doing this implicitly through blocks because this will be confusing for the next person to touch.

But blocks have the advantage that the compiler guarantees that the dropping is implemented correctly rather than having to verify the manual implementation. A comment on the first instance should suffice to explain to the reader what's going on.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs]: I don't know much about Netplan though that seems out of scope for Afterburn. Why not be version agnostic and let the OS deal with version compatibility?

Comment (pothos) [reply src/providers/vmware/amd64.rs]: The OS would run afterburn with this flag and then netplan with a particular version. If any netplan data can come in I thought it's better to narrow the supported values and to error out early - also, the netplan_types crate may not support newer versions.

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs:15]: `guestinfo.userdata` is for cloud-init userdata (and Ignition could have used that, too). The metadata is meant for platform-provided configuration, not user configuration. The name of the guestinfo field and the content is what VMware "specifies" by reusing what cloud-init already supports, and deviating from to something afterburn-specific would defeat the point - one can already do OS-specific things and this was proposed as an OS-independent way of network configuration. It's not so nice that cloud-init defines this in the end but at least netplan is not directly tied to cloud-init. @akutz can maybe give more background.

File: Cargo.toml
Comment (pothos) [reply Cargo.toml]: Thanks, good idea, did this now

File: docs/release-notes.md
Comment (pothos) [reply docs/release-notes.md]: Thanks for the hint

**Comment by pothos** [comment]
> But blocks have the advantage that the compiler guarantees that the dropping is implemented correctly rather than having to verify the manual implementation. A comment on the first instance should suffice to explain to the reader what's going on.

Have changed it now to use blocks.

> One way to avoid having to package this in Fedora is to gate it behind a feature for now. If we do that though, we should also make sure that CI here is still exercising that path.

Have changed it to be behind a feature gate, and the CI will enable all features.

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs]: Ok, have changed it now to use blocks.

File: src/providers/vmware/amd64.rs
Comment (travier) [reply src/providers/vmware/amd64.rs]: From previous comments:
> The distro would be responsible for configuring Netplan to translate to networkd or NetworkManager as desired.

I don't think we should parse the netplan config in Afterburn. We should leave that to another unit in the initrd or the root to do the translation for the desired backend.

**Comment by travier** [comment]
You probably should link to the cloud-init documentation in the docs if we don't have anything more platform generic.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [reply src/providers/vmware/amd64.rs:15]: Ahhh sorry, I missed the `user` vs `meta` distinction. OK, sounds fine to me.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs]:
```suggestion
            .with_context(|| format!("failed to retrieve guestinfo for {}", key))?
```

File: src/providers/vmware/amd64.rs
Comment (jlebon) [reply src/providers/vmware/amd64.rs]: If the OS wants to error out before getting all the way to Netplan generation, integrators can ship a systemd service that runs right `After` Afterburn to check the config version.

Comment (jlebon) [reply src/providers/vmware/amd64.rs]: And actually, it seems like the only requirement on `netplan-types`? So if we drop this function, we should be able to get rid of the feature gating.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [reply src/providers/vmware/amd64.rs:15]: OK right, I totally missed that `guestinfo.metadata` is used today on VMware + cloud-init to pass Netplan configs. That makes a lot more sense now.

File: docs/usage/vmware-netplan-guestinfo-metadata.md
Comment (jlebon) [inline docs/usage/vmware-netplan-guestinfo-metadata.md]: Here and below: we capitalize Afterburn in prose.

File: docs/usage/vmware-netplan-guestinfo-metadata.md
Comment (jlebon) [inline docs/usage/vmware-netplan-guestinfo-metadata.md]: This seems overly specific to how Netplan is actually integrated in the OS. I'd just leave it as e.g.:

> The network environment can vary between VMware servers and instead of leaking these requirements into userdata snippets, a well-known guestinfo metadata field can be used. The guestinfo metadata field is OS-independent and supported by cloud-init and Afterburn. It accepts Netplan configs which the OS can then render down using the backend of choice.

File: docs/usage/vmware-netplan-guestinfo-metadata.md
Comment (jlebon) [inline docs/usage/vmware-netplan-guestinfo-metadata.md]: This type of stuff belongs better in `distro.md` since it's not user-facing but for integrators.

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs]: Sure, I'll drop it (my main motivation for the validation was to follow what usually was done)

Comment (jepio) [reply src/providers/vmware/amd64.rs]: there is one field that it would make sense to parse out of `guestinfo.metadata` and that's `local-hostname`. But we could do that without the `netplan-types` dependency, by carrying our own simple struct definition.

i would suggest doing this as a follow-up, what do you think?

Comment (jlebon) [reply src/providers/vmware/amd64.rs]: Makes sense to me! :+1:

Comment (pothos) [reply src/providers/vmware/amd64.rs]: And we also had the idea that `netplan-types` could be used in the providers for a NetworkManager-capable alternative to the `--network-units=` flag that Flatcar uses. If that needs arises we can still add it and it's not really needed at the moment.

Comment (pothos) [reply src/providers/vmware/amd64.rs]: Found one problem with dropping the parsing: Netplan seems to want YAML but cloud-init is also fine with JSON which the current PR here would convert to YAML. I guess we would have to drop JSON support then. I'll stop for now and leave the feature gate.

**Comment by pothos** [comment]
I looked into removing the `netplan-types` crate but that means dropping JSON support. Cloud-init handles both YAML and JSON, and since @akutz asked for JSON in https://github.com/coreos/afterburn/pull/888#discussion_r1154803578 I don't want to remove it if the feature gate is ok.

File: docs/usage/vmware-netplan-guestinfo-metadata.md
Comment (pothos) [reply docs/usage/vmware-netplan-guestinfo-metadata.md]: Moved, thanks

File: docs/usage/vmware-netplan-guestinfo-metadata.md
Comment (pothos) [reply docs/usage/vmware-netplan-guestinfo-metadata.md]: Thanks, applied

File: docs/usage/vmware-netplan-guestinfo-metadata.md
Comment (pothos) [reply docs/usage/vmware-netplan-guestinfo-metadata.md]: Thanks, done

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs]: `local-hostname` is another top-level entry and not part of `netplan-types` because it's a cloud-init thing. Supporting and parsing that would be done with a separate struct.

**Comment by jlebon** [comment]
Sorry, can you explain the issue in more details? Is the issue that we're not retaining the original format? Can't we note down which format it was in at deserialization time and make sure to reserialize it in the same format?

**Comment by pothos** [comment]
The metadata can be either [JSON or YAML](https://cloudinit.readthedocs.io/en/latest/reference/datasources/vmware.html#guestinfo-keys) but the output for Netplan must be YAML, thus it gets converted from JSON to YAML with the `netplan-types` crate.
I thought that when we drop the crate we drop JSON support but maybe I can also do a generic JSON to YAML conversion with serde but we need to filter unknown keys from the YAML output.

**Comment by jepio** [comment]
> The metadata can be either [JSON or YAML](https://cloudinit.readthedocs.io/en/latest/reference/datasources/vmware.html#guestinfo-keys) but the output for Netplan must be YAML

I thought YAML is a superset of JSON so valid JSON is automatically also valid YAML.

**Comment by pothos** [comment]
Updated it to remove the netplan-types dependency.

> I thought YAML is a superset of JSON so valid JSON is automatically also valid YAML.

Yeah, maybe I delete the match statement.

**Comment by pothos** [comment]
Removed the match statement because it works to only rely on YAML parsing.

File: src/cli/multi.rs
Comment (jlebon) [inline src/cli/multi.rs]: Since we only write down one config, should this be

```suggestion
    #[arg(long = "netplan-config", value_name = "path")]
```

?

File: src/cli/multi.rs
Comment (jlebon) [inline src/cli/multi.rs]: Needs updating

File: src/cli/multi.rs
Comment (jlebon) [inline src/cli/multi.rs]:
```suggestion
            .map_or(Ok(()), |x| metadata.write_netplan_config(x))
```

?

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs]: Let's add error context here.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs]: For clarity, can we reformat this as

```rust
let ... = json!({
	...
})
```

?

I don't _think_ `cargo fmt` will revert it.

Comment (jlebon) [inline src/providers/vmware/amd64.rs]:
```suggestion
            let netplan_config_filtered = json!({ "network": netplan_config_unfiltered.get("network").context("no 'network' key found")?});
```

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs]: Minor: I'd collapse those two.

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs:101]: I think we could avoid some of the indentation here by leveraging pattern matching more. E.g. we could reflow this function as:

```rust
match (
	guestinfo_metadata_encoding_val.as_deref(),
	guestinfo_metadata_raw
) {
	(Some("base64" | "b64")) => {
		...
		Ok(Some(...))
	...
	(Some("") | None, guestinfo_metadata_raw) => Ok(...),
	(Some(enc), _) => Err(anyhow!("unknown guestinfo.metadata.encoding {enc}")),
}
```

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs]: It seems like that's the only reason we're adding `preserve_order`?

Can we avoid this by slightly tweaking the tests to just not have multiple subkeys in a key? We're not actually semantically parsing this anyway, and we're still testing the main thing, which is that we can parse JSON and YAML and exclude unknown top-level fields.

File: src/cli/multi.rs
Comment (pothos) [reply src/cli/multi.rs]: My idea was that the Equinix Metal provider could also offer Netplan instead of networkd directly but yes, since it works to write one large config file instead of many being merged, it would still be ok in the singular even then.

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs]: Yes, done

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs:101]: Thanks

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs]: Yes, works

File: src/providers/vmware/amd64.rs
Comment (jlebon) [inline src/providers/vmware/amd64.rs]: Would be nice to the user to print the encoding string found in the error.

**Review by jlebon — APPROVED** [review]
One minor comment, but LGTM overall. Nice work!

File: src/providers/vmware/amd64.rs
Comment (pothos) [reply src/providers/vmware/amd64.rs]: Sure, done

**Review by jlebon — APPROVED** [review]

**Pull request merged**
